nav2-usability/scenario-code/dynamic-linking/router.dart

```diff
@@ -0,0 +1,233 @@
+// Copyright 2021, the Flutter project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:math';
+
+import 'package:flutter/material.dart';
+
+void main() {
+  runApp(WishlistApp());
+}
+
+class Wishlist {
+  final String id;
+
+  Wishlist(this.id);
+}
+
+class AppState {
+  final List<Wishlist> wishlists = <Wishlist>[];
+  Wishlist? selectedWishlist;
+}
+
+class WishlistApp extends StatefulWidget {
+  @override
+  _WishlistAppState createState() => _WishlistAppState();
+}
+
+class _WishlistAppState extends State<WishlistApp> {
+  final BookRouterDelegate _routerDelegate = BookRouterDelegate();
+  final BookRouteInformationParser _routeInformationParser =
+  BookRouteInformationParser();
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      title: 'Books App',
+      routerDelegate: _routerDelegate,
+      routeInformationParser: _routeInformationParser,
+    );
+  }
+}
+
+class BookRouteInformationParser extends RouteInformationParser<AppRoutePath> {
+  @override
+  Future<AppRoutePath> parseRouteInformation(
+      RouteInformation routeInformation) async {
+    final uri = Uri.parse(routeInformation.location!);
+    // Handle '/'
+    if (uri.pathSegments.length == 0) {
+      return AppRoutePath();
+    }
+
+    // Handle '/wishlist/:id'
+    if (uri.pathSegments.length == 2) {
+      if (uri.pathSegments[0] != 'wishlist') return AppRoutePath();
+      final id = uri.pathSegments[1];
+      return AppRoutePath(id: id);
+    }
+
+    // Handle unknown routes
+    return AppRoutePath();
+  }
+
+  @override
+  RouteInformation restoreRouteInformation(AppRoutePath path) {
+    late final String location;
+    if (path.id == null) {
+      location = '/';
+    } else {
+      location = '/wishlist/${path.id}';
+    }
+    return RouteInformation(location: location);
+  }
+}
+
+class BookRouterDelegate extends RouterDelegate<AppRoutePath>
+    with ChangeNotifier, PopNavigatorRouterDelegateMixin<AppRoutePath> {
+  final GlobalKey<NavigatorState> navigatorKey;
+
+  AppState appState = AppState();
```

Comment:
this can be a change notifier that owns by _WishlistAppState, which passes into BookRouterDelegate. BookRouterDelegate can listen to the change notifier and call its notifylisteners

Reply:
👍

Reply:
Change notifier needs to be disposed.
If BookRouterDelegate owns this state, is should also be in charge of disposing it in its dispose method.

Reply:
Fixed